src/Microsoft.Restier.Core/ApiBase.cs

```diff
@@ -137,6 +131,31 @@ public void Dispose()
         }
 
         /// <summary>
+        /// Configure the entire service set for this API. Descendants can override this method to choose
+        /// from the default service sets provided by RESTier.
+        /// </summary>
+        /// <param name="apiType">
+        /// CLR type of this <see cref="ApiBase"/>.
+        /// </param>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection"/> with which to create an <see cref="ApiConfiguration"/>.
+        /// </param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [CLSCompliant(false)]
+        protected virtual IServiceCollection ConfigureApiServices(Type apiType, IServiceCollection services)
+        {
+            var customizer = ApiConfiguration.Customize(apiType);
+            services.Apply(customizer.InnerMost);
+
+            return this.ConfigureApi(services)
+                .Apply(customizer.PrivateApi)
+                .UseAttributes(apiType)
+                .UseConventions(apiType)
+                .Apply(customizer.Overrides)
+                .Apply(customizer.OuterMost);
+        }
+
```

Comment:
1.  Two notes
2. Ray's pull request......


thanks for your response, I think more about this and have a short discussion with Lewis,

The valid requests now,
1. Allow register a model builder before conversion model build to build entity set. And this could between Conversion model build and EF model builder to add more entity set.
2. Allow to register as a head model build so user can log a whole model.

Also behaviors are depended in some cases:
1. Inner most register may take no effect if same service we register later use cutoff.
2. Even user register as out most, it could be consider like inner most if user write code like 
inner.call()
customizedlogic();
3. Even in our code, when we call inner to do something, we may call it first or last.
Like model builder, we call inner first.
For QueryExecutor, with some check, and call inner later.
4. User can extends current service class, control whole logic. Then register out most without inner.

Any comments of these statements?

Based on these statements,
1. Inner most is not a concern as user can register out most and call inner first.

2. We must allow user to register out most, and I think we can achieve via 
base.Configure()
service.cutoff()
service.chainprevious()

3. We need to allow use to register after conversion (and before DBContext), user can do like 
service.AddCoreService() // Includes attributes and other registered services.
service.UseDbContext()// Optional
service.ChainPrevious()// User customized service
service.UserConversions()
service.ChainPrevious()// User customized service

I did not use other chains like inner/out/private, as I think it is hard for end user to understand what services are registered between each of these service chain.

4. For DbAPi, the key is query executor, its configuration will be like 
service.UseInternalService() // Includes attributes and other registered services.
serivice.UseDbContext()
service.UserConversions()

5. For RESTierQueryExecutor, a new extension method in WebApi project?

We can add more document of ConfigAPi and each user methods (maybe user change to AddEFService or something similar).
For internal service, 
And we allow user to have full control of the chain.